src/java/org/apache/cassandra/index/sai/disk/PrimaryKeyMapIterator.java

```diff
@@ -0,0 +1,91 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.index.sai.disk;
+
+import java.io.IOException;
+
+import org.apache.cassandra.db.PartitionPosition;
+import org.apache.cassandra.dht.AbstractBounds;
+import org.apache.cassandra.dht.Token;
+import org.apache.cassandra.index.sai.SSTableContext;
+import org.apache.cassandra.index.sai.utils.PrimaryKey;
+import org.apache.cassandra.index.sai.utils.RangeIterator;
+
+/**
+ * Iterates keys in the PrimaryKeyMap of an SSTable.
+ * Iterating keys in the primary key map is faster than reading them from the SSTable data component
+ * because we don't deserialize all the other columns except the primary key columns.
+ * The primary key map components are also likely much smaller than the whole SSTable data component.
+ * <p>
+ * The keys are returned in token-clustering order.
+ */
+public final class PrimaryKeyMapIterator extends RangeIterator
+{
+    private final PrimaryKeyMap keys;
+    private long currentRowId;
+
+    private PrimaryKeyMapIterator(PrimaryKeyMap keys, PrimaryKey min, PrimaryKey max)
+    {
+        super(min, max, keys.count());
+        this.keys = keys;
+        this.currentRowId = keys.exactRowIdForPrimaryKey(min);
+    }
+
+    public static PrimaryKeyMapIterator create(SSTableContext ctx, AbstractBounds<PartitionPosition> keyRange) throws IOException
+    {
+        PrimaryKeyMap keys = ctx.primaryKeyMapFactory.newPerSSTablePrimaryKeyMap();
+        long count = keys.count();
+
+        PrimaryKey.Factory pkFactory = ctx.indexDescriptor.primaryKeyFactory;
+        Token minToken = keyRange.left.getToken();
+        Token maxToken = keyRange.right.getToken();
+        PrimaryKey minKeyBound = pkFactory.createTokenOnly(minToken);
+        PrimaryKey maxKeyBound = pkFactory.createTokenOnly(maxToken);
+        PrimaryKey sstableMinKey = count > 0 ? keys.primaryKeyFromRowId(0) : null;
+        PrimaryKey sstableMaxKey = count > 0 ? keys.primaryKeyFromRowId(count - 1) : null;
+        PrimaryKey minKey = (sstableMinKey == null || minKeyBound.compareTo(sstableMinKey) > 0)
+                            ? minKeyBound
+                            : sstableMinKey;
+        PrimaryKey maxKey = (sstableMaxKey == null || !maxToken.isMinimum() && maxKeyBound.compareTo(sstableMaxKey) < 0)
+                            ? maxKeyBound
+                            : sstableMaxKey;
+        return new PrimaryKeyMapIterator(keys, minKey, maxKey);
+    }
+
+    @Override
+    protected void performSkipTo(PrimaryKey nextKey)
+    {
+        this.currentRowId = keys.exactRowIdForPrimaryKey(nextKey);
```

Comment:
I think ceiling gives us the right semantics to get the row id associated with the key greater than or equal to `nextKey`.

```suggestion
        this.currentRowId = keys.ceiling(nextKey);
```

Similar to another comment I left, we'll need to update `computeNext` to handle the case that `nextKey` is not in the map.